cirq-core/cirq/circuits/circuit.py

```diff
@@ -139,7 +153,10 @@ def freeze(self) -> 'cirq.FrozenCircuit':
 
         if isinstance(self, FrozenCircuit):
             return self
-        return FrozenCircuit(self, strategy=InsertStrategy.EARLIEST, device=self.device)
+
+        if self._device == cirq.UNCONSTRAINED_DEVICE:
+            return FrozenCircuit(self, strategy=InsertStrategy.EARLIEST)
+        return FrozenCircuit(self, strategy=InsertStrategy.EARLIEST, device=self._device)
```

Comment:
The reason I opted for this patter instead of something like:

```python3
ret_device = FrozenCircuit(self, strategy=....)
ret_device._device = self._device
return ret_device
```
Is because the constructor with the device argument makes non trivial behavior changes to the constructor and so that's why I opted to use this pattern in places with deprecations. Removing them once the cycle is over should be relatively easy to do since all we need to do is keep the snippet inside the if statement and delete the else block.